backend-tests/tests/test_inventory.py

```diff
@@ -321,3 +293,278 @@ def test_fail_no_attr_value(self, user):
+                    ],
+                },
+                "status_code": 200,
+                "response": [
+                    {
+                        "id": dev.id,
+                        "attributes": self.dict_to_inventoryattrs(
+                            dev.inventory, scope="inventory"
+                        ),
+                    }
+                    # The following is just the python expression of the
+                    # above operation.
+                    for dev in sorted(
+                        filter(
+                            lambda dev: dev.inventory["version"] != "v3.0",
+                            tenant.devices,
+                        ),
+                        key=lambda dev: json.dumps(dev.inventory["idx"]),
+                        reverse=True,
+                    )
+                ],
+            },
+            {
+                "name": "Error - missing type parameter",
+                "request": {
+                    "filters": [
+                        {
+                            "attribute": "version",
+                            "value": ["v1.0"],
+                            "scope": "inventory",
+                        },
+                    ],
+                },
+                "status_code": 400,
+            },
+            {
+                "name": "Error - valid mongo query unsupported operation",
+                "request": {
+                    "filters": [
+                        {
+                            "type": "$type",
+                            "attribute": "version",
+                            "value": ["int", "string", "array"],
+                            "scope": "inventory",
+                        },
+                    ],
+                },
+                "status_code": 400,
+            },
+        ]
+
+        for test_case in test_cases:
+            self.logger.info("Running test case: %s" % test_case["name"])
+            rsp = invm_v2.with_auth(tenant.api_token).call(
+                "POST", inventory_v2.URL_SEARCH, test_case["request"]
+            )
+            assert rsp.status_code == test_case["status_code"], (
+                "Unexpected status code (%d) from /filters/search response: %s"
+                % (rsp.status_code, rsp.text)
+            )
+
+            if rsp.status_code == 200 and "response" in test_case:
+                body = rsp.json()
+                if body is None:
+                    body = []
+                assert len(body) == len(test_case["response"]), (
+                    "Unexpected number of results: %s != %s"
+                    % (test_case["response"], body)
+                )
+
+                if len(body) > 0:
+                    for dev in test_case["response"]:
+                        for api_dev in body:
+                            if dev["id"] == api_dev["id"]:
+                                assert_device_attributes(dev, api_dev)
```

Comment:
We need to assert that all `dev` devices are present in `api_dev`.

If for some strange reason you get the same number of results but no match in any of the devices, then this would result in a false positive.

Reply:
fixed